### ISSUE TITLE
BAM and reported generated with no contents; (ERR): bowtie2-align died with signal 13 (PIPE)

### ISSUE THREAD

**Issue by MichaelMW** [FelixKrueger/Bismark#65]
Summary:
BAM and reported generated with no contents, i.e.:

2.1M Oct  6 11:00 test_data.fastq_C_to_T.fastq
0 Oct  6 11:01 test_data_bismark_bt2.bam
0 Oct  6 11:01 test_data_bismark_bt2_SE_report.txt

Tried bowtie1, bowtie2.2.5, bowtie2.2.9; Also tried various genomes. 

For bowtie2.2.9 and 2.2.5, the log reads:
"First alignment found" ...
"writing bisulfite mapping results to xxx.sam" ...
"Reading in the sequence file test_data.fastq" ... 
"(ERR): bowtie2-align died with signal 13 (PIPE)"

For bowtie1:
"Reading in the sequence file test_data.fastq
Error while flushing and closing output
Error while flushing and closing output
terminate called after throwing an instance of 'int'
terminate called after throwing an instance of 'int'"

System Info:
- Linux 3.10.0-327.4.4.el7.x86_64
- Intel(R) Xeon(R) CPU E3-1226 v3 @ 3.30GHz
- 20GB RAM

Steps (for bowtie2.2.9)
- 1. Download GRCh38 from 
ftp://ftp.ensembl.org/pub/release-86/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
- 2. prepare genome 
bismark_genome_preparation --path_to_bowtie ~/tools/bowtie2-2.2.9/ --verbose /home/mengchi/database/genomes/forBismark/GRCh38 &> prepareGenome.bt2.2.9.txt
- 3. running bismark
bismark --score_min L,0,-0.6 /home/mengchi/database/genomes/forBismark/GRCh38 test_data.fastq &> runningBismark.bt2.2.9.txt

Here attached the detailed logs:
[runningBismark.bt2.2.9.txt](https://github.com/FelixKrueger/Bismark/files/514501/runningBismark.bt2.2.9.txt)
[prepareGenome.bt2.2.9.txt](https://github.com/FelixKrueger/Bismark/files/514500/prepareGenome.bt2.2.9.txt)

Also attached is the log for running bismark with bowtie1. 
[runningBismark.bt1.txt](https://github.com/FelixKrueger/Bismark/files/514509/runningBismark.bt1.txt)

Thank you in advance!

**Comment by FelixKrueger** [comment]
Hi Michael,

`Died with signal 13 (PIPE)` usually means that you ran out of memory and so the Bowtie instances crash. I can check again tomorrow but I  would assume that the genome you downloaded is just about too large for your computer to handle, is that a possibility? Were you running anything else at the time?

**Comment by MichaelMW** [comment]
Thank you for the swift response. 
So I tried another run. To ensure minimal memory requirement, I prepared a short mock genome.
Indeed, I'm no longer getting the `Died with signal 13 (PIPE)` error. 
However, the resulting bam and report is still 0 in size. 
The run time is within a minute, with minimal CPU and memory usage. 

```
2.1M Oct  6 22:25 test_data.fastq_C_to_T.fastq
0 Oct  6 22:25 test_data_bismark_bt2.bam
0 Oct  6 22:25 test_data_bismark_bt2_SE_report.txt
```

Here are my code:
```
bismark_genome_preparation --path_to_bowtie ~/tools/bowtie2-2.2.9/ --verbose ref &> prepareGenome.mock.txt
bismark --score_min L,0,-0.6 ref test_data.fastq  &> runningBismark.mock.txt
```

And logs:

[prepareGenome.mock.txt](https://github.com/FelixKrueger/Bismark/files/515325/prepareGenome.mock.txt)
[runningBismark.mock.txt](https://github.com/FelixKrueger/Bismark/files/515324/runningBismark.mock.txt)


Thanks again.

**Comment by FelixKrueger** [comment]
This is very odd, it seems to convert the FastQ files, find the first alignments, then process 10000 lines (i.e. all sequences) but then doesn't produce any output whatsoever? At the very least it should print the report on-screen, unless it hangs and stalls just before it gets to that section. Also, the BAM file should not be empty, it should at least contain the header lines and the few alignments you got.

So I am wondering whether it is just something weird with line-endings on the test file (can you do `file test_data.fastq`?) that need to be converted to Linux (e.g. `dos2unix test_data.fastq` or `mac2unix test_data.fastq`). Can you do a `which samtools` to see if which version of Samtools you are using? Can you read and write to BAM files stand-alone? Can you also run the command without redirecting the output to a file? Does it complete at all (I can't seem to think of a case why Bismark would ever not complete or die without some kind of error message)? Cheers, Felix

**Comment by MichaelMW** [comment]
Fixed! You are pointing to the right direction. The problem is with samtools, which failed to run due to lack of a dependency called ncurses in anaconda. This problem is fixed by `conda install -c anaconda ncurses=5.9`

This is no fault of Bismark. Regardless, to improve the robustness, a check on validity of `samtools --version` can be added at the beginning of bismark. 

Cheers, Michael

**Comment by FelixKrueger** [comment]
Hi Michael, I'm glad this is fixed now. I'll definitely keep in mind to improve the Samtools check for a future version.
